jax/_src/numpy/fft.py

```diff
@@ -55,7 +58,20 @@ def _fft_core(func_name, fft_type, a, s, axes, norm):
     axes = tuple(range(a.ndim - len(axes), a.ndim))
     a = jnp.moveaxis(a, orig_axes, axes)
 
-  if s is None:
+  if s is not None:
+    a = jnp.asarray(a)
+    in_s = list(a.shape)
+    for axis, x in zip(axes, s):
+      in_s[axis] = x
+    if fft_type == xla_client.FftType.IRFFT:
+      in_s[-1] = (in_s[-1] // 2 + 1)
+
+    if any([x != y for x, y in zip(in_s, a.shape)]):
+      # Cropping
+      a = a[tuple(map(slice, in_s))]
+      # Padding
+      a = lax.pad(a, lax._const(a, 0), [(0, x-y, 0) for x, y in zip(in_s, a.shape)])
+  else:
     if fft_type == xla_client.FftType.IRFFT:
```

Comment:
I wonder if we could reuse this logic for building up `in_s`, to avoid duplicated code?

E.g.,
```python
    if fft_type == xla_client.FftType.IRFFT:
      in_s = [a.shape[axis] for axis in axes[:-1]]
      if axes:
        in_s += [max(0, 2 * (a.shape[axes[-1]] - 1))]
    else:
      in_s = [a.shape[axis] for axis in axes]

    if s is None:
      s = in_s
    else:
      a = a[tuple(map(slice, in_s))]
      a = jnp.pad(a, [(0, x-y) for x, y in zip(in_s, a.shape)])
```

Reply:
It would be nice if this could be simplified. However `in_s` is different because it represents the full input shape, while s and axes could present less axes. E.g. if `a.shape = [5, 3, 4, 5]`, `s = [10]`, and `axes = [1]` then we have `in_s=[5,10,4,5]`.

Since `lax.pad`/`jnp.pad` require all axes to be supplied, I think it's necessary for `in_s` to contain all even if no padding/cropping is needed.

Reply:
Ah, good point.